src/adapters/ip_camera_adapter/mod.rs

```diff
@@ -0,0 +1,304 @@
+        Ok(response)
+    }
+}
+
+impl<T: Controller> Service for IpCameraService<T> {
+    fn get_properties(&self) -> ServiceProperties {
+        self.properties.clone()
+    }
+
+    // Starts the service, it will just spawn a thread and send messages once
+    // in a while.
+    fn start(&self) {
+        let props = self.properties.clone();
+        let controller = self.controller.clone();
+
+        info!("Starting IpCameraService for ID: {} IP: {} dir: {}", props.id, self.ip, self.snapshot_dir);
+        match fs::create_dir_all(&self.snapshot_dir) {
+            Ok(_) => {},
+            Err(err) => {
+                error!("Unable to create directory {}: {}", self.snapshot_dir, err);
+                controller.remove_service(props.id.to_string());
+            }
+        }
+    }
+
+    fn stop(&self) {
+        let props = self.properties.clone();
+        info!("Stopping IpCameraService for ID: {} IP: {}", props.id, self.ip);
+    }
+
+    // Processes a http request.
+    fn process_request(&self, req: &mut Request) -> IronResult<Response> {
+        let cmd = req.extensions.get::<Router>().unwrap().find("command").unwrap_or("");
+        info!("IpCameraAdapter {:?} received command {} for camera ip {}", req, cmd, self.ip);
+
+        match cmd {
+            "snapshot" => return self.cmd_snapshot(),
+            "list" => return self.cmd_list(),
+            "get" => {
+                if let Some(filename) = req.url.query.clone() {
+                    // For now we assume that the query is the filename
+                    return self.cmd_get(filename);
+                }
+                let mut response = Response::with("get needs a query");
+                response.status = Some(Status::BadRequest);
+                response.headers.set(ContentType::plaintext());
+                return Ok(response);
+            },
+            _ => {}
+        }
+
+        let mut response = Response::with(format!("Unrecognized command: {}", cmd));
+        response.status = Some(Status::BadRequest);
+        response.headers.set(ContentType::plaintext());
+        Ok(response)
+    }
+}
+
+pub struct IpCameraAdapter<T> {
+    name: String,
+    controller: T
+}
+
+impl<T: Controller> IpCameraAdapter<T> {
+    pub fn new(controller: T) -> Self {
+        debug!("Creating IpCameraAdapter");
+        IpCameraAdapter { name: "IpCameraAdapter".to_owned(),
+                       controller: controller }
+    }
+}
+
+impl<T: Controller> ServiceAdapter for IpCameraAdapter<T> {
+    fn get_name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn start(&self) {
+        //let mut id = 0;
+        let controller = self.controller.clone();
+        thread::spawn(move || {
+            controller.adapter_started("IpCamera Service Adapter".to_owned());
+
+            // TODO: We should do discovery and instantiate a new IpCameraService
+            //       for each camera that we discover.
+
```

Comment:
@aosmond has been working in upnp, which the DLink cameras supposedly support. So I'm going to look into that. I should be able to get a bunch of information from each camera (name, IP address, model, serial number).